crates/bevy_ecs/src/bundle.rs

```diff
@@ -89,44 +84,69 @@ pub unsafe trait Bundle: Send + Sync + 'static {
     where
         Self: Sized;
 
-    /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This will
-    /// [`std::mem::forget`] the bundle fields, so callers are responsible for dropping the fields
-    /// if that is desirable.
+    /// Calls `func` on each value, in the order of this bundle's Components. This will
+    /// give `func` ownership of the bundle fields.
+    ///
+    /// # Safety:
+    /// Must not use the component value after `func` is called with it
     fn get_components(self, func: impl FnMut(*mut u8));
 }
 
 macro_rules! tuple_impl {
     ($($name: ident),*) => {
-        unsafe impl<$($name: Component),*> Bundle for ($($name,)*) {
-            #[allow(unused_variables)]
+        // SAFE: Component is returned in tuple-order. [Bundle::from_components] and [Bundle::get_components] use tuple-order
+        unsafe impl<$($name: Bundle),*> Bundle for ($($name,)*) {
+            #[allow(unused_variables, unused_mut)]
             fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId> {
-                vec![$(components.init_component::<$name>(storages)),*]
+                let mut res = vec![];
+                $(
+                    res.extend(<$name as Bundle>::component_ids( components, storages));
+                )*
+                res
             }
 
             #[allow(unused_variables, unused_mut)]
             #[allow(clippy::unused_unit)]
             unsafe fn from_components(mut func: impl FnMut() -> *mut u8) -> Self {
-                #[allow(non_snake_case)]
-                let ($(mut $name,)*) = (
-                    $(func().cast::<$name>(),)*
-                );
-                ($($name.read(),)*)
+                ($(
+                    $name::from_components(&mut func),
+                )*)
             }
 
             #[allow(unused_variables, unused_mut)]
             fn get_components(self, mut func: impl FnMut(*mut u8)) {
                 #[allow(non_snake_case)]
                 let ($(mut $name,)*) = self;
                 $(
-                    func((&mut $name as *mut $name).cast::<u8>());
-                    std::mem::forget($name);
+                    $name.get_components(&mut func);
                 )*
             }
         }
     }
 }
 
-all_tuples!(tuple_impl, 0, 15, C);
+all_tuples!(tuple_impl, 0, 15, B);
+
+/// Implement `Bundle` for all `Component`s, as any component is trivially a bundle of itself
+// SAFE: `component_ids` returns the correct ComponentId, which is the same type as given in `get_components` and
+// taken in `from_components`
+unsafe impl<T: Component> Bundle for T {
+    fn component_ids(components: &mut Components, storages: &mut Storages) -> Vec<ComponentId> {
+        vec![components.init_component::<T>(storages)]
```

Comment:
It's possible. This is a pre-existing issue with the `Bundle` api, which should be fixed seperately.

Reply:
I've heard somewhere that allocation (possibly specifically through the system allocator) is special in llvm, which lets it get optimised out sometimes; since llvm knows the contract of allocation.

Reply:
The function would just add to an existing Vec, by taking it as a `&mut Vec<ComponentId>`

Reply:
> I've heard somewhere that allocation (possibly specifically through the system allocator) is special in llvm, which lets it get optimised out sometimes; since llvm knows the contract of allocation.

True. This applies to the system allocator and any `#[global_allocator]` (the later only when using rust's fork of llvm). I believe it only happens in ideal circumstances though.

Reply:
It turns out that this PR makes it significantly easier to actually fix this, so I've merged that change in.